src/src/PluginStructs/P164_data_struct.cpp

```diff
@@ -0,0 +1,889 @@
+#define ENS160_OPMODE_RESET         0xF0    // RESET 
+#define ENS160_OPMODE_DEEP_SLEEP    0x00    // DEEPSLEEP
+#define ENS160_OPMODE_IDLE          0x01    // IDLE
+#define ENS160_OPMODE_STD           0x02    // STANDARD
+#define ENS160_OPMODE_LP            0x03    // LOW POWER (ENS161 only)
+#define ENS160_OPMODE_ULP           0x04    // ULTRA LOW POWER (ENS161 only)
+#define ENS160_OPMODE_CUSTOM        0xC0    // Not specified in datasheet
+
+// ENS160 undefined bitfields?
+#define ENS160_BL_CMD_START         0x02
+#define ENS160_BL_CMD_ERASE_APP     0x04
+#define ENS160_BL_CMD_ERASE_BLINE   0x06
+#define ENS160_BL_CMD_WRITE         0x08
+#define ENS160_BL_CMD_VERIFY        0x0A
+#define ENS160_BL_CMD_GET_BLVER     0x0C
+#define ENS160_BL_CMD_GET_APPVER    0x0E
+#define ENS160_BL_CMD_EXITBL        0x12
+
+// ENS160 undefined bitfields?
+#define ENS160_SEQ_ACK_NOTCOMPLETE  0x80
+#define ENS160_SEQ_ACK_COMPLETE     0xC0
+
+#define IS_ENS160_SEQ_ACK_NOT_COMPLETE(x)   (ENS160_SEQ_ACK_NOTCOMPLETE == (ENS160_SEQ_ACK_NOTCOMPLETE & (x)))
+#define IS_ENS160_SEQ_ACK_COMPLETE(x)       (ENS160_SEQ_ACK_COMPLETE == (ENS160_SEQ_ACK_COMPLETE & (x)))
+
+// ENS160 STATUS bitfields
+#define ENS160_STATUS_STATAS        0x80    // STATAS: Indicates that an OPMODE is rumming
+#define ENS160_STATUS_STATER        0x40    // STATER: High indicated that an error is detected
+#define ENS160_STATUS_VALIDITY      0x0C    // VALIDITY FLAG
+#define ENS160_STATUS_NEWDAT        0x02    // NEWDAT: 1= New data in data registers available
+#define ENS160_STATUS_NEWGPR        0x01    // NEWGRP: 1= New data in GRP_READ registers available 
+
+// Checkers for bitfields in STATUS register
+#define IS_NEWDAT(x)                (ENS160_STATUS_NEWDAT == (ENS160_STATUS_NEWDAT & (x)))
+#define IS_NEWGPR(x)                (ENS160_STATUS_NEWGPR == (ENS160_STATUS_NEWGPR & (x)))
+#define IS_NEW_DATA_AVAILABLE(x)    (0 != ((ENS160_STATUS_NEWDAT | ENS160_STATUS_NEWGPR ) & (x)))
+#define GET_STATUS_VALIDITY(x)      (((x) & ENS160_STATUS_VALIDITY) >> 2)
+
+// TODO: add comment on this
+#define CONVERT_RS_RAW2OHMS_I(x)    (1 << ((x) >> 11))
+#define CONVERT_RS_RAW2OHMS_F(x)    (pow (2, (float)(x) / 2048))
+
+// Form IDs used on the device setup page. Should be a short unique string.
+#define P164_GUID_TEMP_T     "f08"
+#define P164_GUID_TEMP_V     "f09"
+#define P164_GUID_HUM_T      "f10"
+#define P164_GUID_HUM_V      "f11"
+
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+// Constructor                                                                                   //
+///////////////////////////////////////////////////////////////////////////////////////////////////
+P164_data_struct::P164_data_struct(struct EventStruct *event) :
+  i2cAddress(P164_I2C_ADDR)
+{
+  #ifdef P164_ENS160_DEBUG
+    Serial.println(F("ENS160: Constructor"));
+    Serial.print("ENS160: I2C address =");
+    Serial.println(i2cAddress, HEX);
```

Comment:
Please use `addLog()` or `addLogMove()` for logging statements, and also guard those with `if (logLevelActiveFor(LOG_LEVEL_nnnn))`, so all logging can also be viewed via the web interface.
Also use the (relatively new) `concat()` and `strformat()` functions for concatenating / printf-like string building for resource reduction.